mitodl/ocw-to-hugo
# Fix external links

#### Pre-Flight checklist


#### What are the relevant tickets?
Fixes #188 

#### What's this PR do?
Fixes a couple issues relating to relative URLs:
 - course links were being resolved to the current course instead of the linked course
 - PDFs and other files in other courses referenced with relative URLs should work, with some exceptions described by #180 

Major changes: 
 - `pathLookup` is updated to keep track of some other useful information besides just the uid, course, and path. This allows `resolveUidForLink` enough information to resolve relative links between courses and to sections within those courses. `pathLookup` is also split into `byUid` and `byCourse` to allow lookups by both (objects are shared to save memory). 
   - Both `resolveUidForLink` and `resolveRelativeLink` make use of `pathLookup` and no longer look into the course JSON for information about PDFs or sections. 
 - `makePdfLink` is created from code shared by `resolveUidForLink` and `resolveRelativeLink`
 - `resolveRelativeLink` checks for `/courses` before processing a link. This means `/resource` links are no longer resolved (I created #192 for this) but it was being parsed incorrectly anyway
 - For relative links to pdfs, both the filename and the parent path must match. This fixes some incorrect links where a course had two PDFs with the same name and the wrong one was chosen.
 - The logic for handling course and section links should be fixed so you no longer have `/sections` for a course link

#### How should this be manually tested?
Convert all courses and do a directory comparison between the master branch and this one. You should see many changes, probably too many to check all of them. In general you should see:
 - Course links should change to point correctly to `/courses/course-id-here/` or possibly `{{< baseurl >}}/` for links to the same course. Section and subsection links to other courses should also be rewritten correctly.
 - 1-103 under the syllabus has a few changed ZIP file links. I could not verify A1 because the link is broken on the legacy system. The B1 link is changed but it looks like that is fixed.
 - External PDFs links are fixed. For example see 11-800 under readings (this is the bug from #188)
 - Some lookups of other static assets are also fixed. For example see the assignments page on `6-864-advanced-natural-language-processing-fall-2005`. The first link was matching `poscounts.gz` instead of `counts.gz`, but now it matches with the right file.
